compiler/rustc_target/src/abi/call/aarch64.rs

```diff
@@ -30,7 +30,6 @@ where
     C: HasDataLayout,
 {
     if !ret.layout.is_aggregate() {
-        ret.extend_integer_width_to(32);
```

Comment:
Do we have special handling for Apple platforms? As I pointed out in https://github.com/rust-lang/rust/issues/97463#issuecomment-1144995863, Apple uses a [non-standard calling convention](https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms#:~:text=The%20caller%20of%20a%20function%20is%20responsible%20for%20signing%20or%20zero%2Dextending%20any%20argument%20with%20fewer%20than%2032%20bits.%20The%20standard%20ABI%20expects%20the%20callee%20to%20sign%20or%20zero%2Dextend%20those%20arguments.) so that this extension is still needed.

Reply:
I'll look into this.

Reply:
(also, the use of `32` there is almost certainly a cut-and-paste bug, as pointed out by @spastorino in yesterday's T-compiler meeting.)

Reply:
No, 32 is correct. The Apple doc says "The caller of a function is responsible for signing or zero-extending any argument with fewer than 32 bits".

Reply:
hmm, okay; the use of 32 is not as "obviously wrong" as I had thought.

Some questions that came up for me while looking further into this:
 * As noted by @HeroicKatora in their [comment on the issue](https://github.com/rust-lang/rust/issues/97463#issuecomment-1140210248), it is unclear whether one is supposed to apply those rules to the return values in same manner as the function arguments
 * I am reading and re-reading the section of the Windows ABI doc (supposedly excerpted directly from ARM's docs) on ["Prepadding and extension of arguments"](https://docs.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions?view=msvc-170#stage-b--pre-padding-and-extension-of-arguments) and the subsequent ["Assignment of arguments to registers and stack"](https://docs.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions?view=msvc-170#stage-c--assignment-of-arguments-to-registers-and-stack); from the titles I had expected to see a bullet specifying at least one case where (sign-or-zero) extension occurs, but I think it pretty clearly indicates that each value is copied into the least significant bits of the corresponding register or stack, and the remaining bits are left as unspecified values.
 * In the section dedicated to [return values](https://docs.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions?view=msvc-170#return-values), it doesn't say anything about sign or zero extension...

Reply:
As for the return value, https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#65result-return says:
> If the type, T, of the result of a function is such that
> ```c
> void func(T arg)
> ```
> would require that arg be passed as a value in a register (or set of registers) according to the rules in [Parameter passing](https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#parameter-passing), then the result is returned in the same registers as would be used for such an argument.

So the same rule should apply regardless parameter or return value.

So using a language lawyer reasoning, if Apple adds a rule to the parameter passing then the standard AArch64 rule that mandates parameter and return value be passed the same way kicks in and adds that rule to the return value too. This is consistent with the behaviour of clang. I agree that Apple should be explicit about this though.

For the Windows ARM64 ABI doc, I suppose it lacks mentioning about sign-extension or zero-extension because there're no such requirements. If you look at the ARM ABI doc it does mention about such extensions.

Reply:
okay thanks for spelling this all out for me, it saved me quite a bit of time trying to double check it all myself.